woocommerce/google-listings-and-ads
# Migrate jest tests to use Node.js 20 and make components compatible with upgraded packages

### Changes proposed in this Pull Request:

Closes #1098

To make this repo use Node.js 20, this PR migrates jest tests along with two rule fixes for eslint:

- Adjust the settings so that jest can run:
   - Fix that `is-plain-obj@4.1.0` doesn't provide the CommonJS build.
   - Force `uuid` to resolve with the CommonJS entry point as jest doesn't support `package.json.exports`.
   - Fix `@woocommerce/components` still using incompatible `@woocommerce/currency`.
   - Ignore that `qrcode.react` module is incorrectly listed in dev dependencies of `@automattic/components`.
   - Fix the React versioning conflicts between `@wordpress/*` and `@woocommerce/*`.
- Update timer mocks to deprecate the `timers` option.
- Migrate test cases to `@testing-library/user-event` v14.
   - Invoke `userEvent.setup()` in the test itself and before the component is rendered.
   - Adding `async` and `await` as all APIs are asynchronous.
   - No need to be wrapped in `act`.
   - Replace the simulation of pressing buttons with the `keyboard` API.
   - `clear()` is changed to only support editable elements only.
- Prevent form submission during testing.
- The ARIA role of `<img>` is considered as `presentation` if its `alt` is an empty string.
- Update snapshots to reflect the format changes.
- Fix `CampaignNameCell` tests by replacing `fireEvent` with `userEvent`.
- Resolve eslint errors for the `testing-library/no-wait-for-multiple-assertions` rule.
- Resolve eslint errors for the `testing-library/no-unnecessary-act` rule.
- Ignore the deprecation warnings of `defaultProps` from React by calling `console.error()`.
- Remove the `structuredClone` polyfill no longer needed.

In addition, it also needs a few adjustments to components or their tests to be compatible with upgraded `@wordpress/components` and `@woocommerce/components`:

- Prevent the `<Dashboard>` tests from calling `window.fetch`.
- Adjust `<Stepper>` related tests to wait for stepper content to be rendered.
- Make `<AppTooltip>` compatible with the `placement` prop introduced by WordPress 6.4 to replace `position`.
- Rewrite the replacer of `JSON.stringify` in `getReportKey` to ensure it returns the same key regardless of the query keys' order.
- Replace deprecated `event.keyCode` with `event.code` for `<VerificationCodeControl>`.
- Avoid accidentally using the event object to reset the asset group in `<CampaignAssetsForm>`.

#### 💡 The logical correlation between commits is low and highly independent, so it would be easier to review changes by looking at each commit individually. 🙏 

### Detailed test instructions:

1. Set Node.js to v20
2. Run `npm install` to see if it can finish without errors
3. Check if the following scripts work well
   - `npm run lint:js`
   - `npm run test:js`
4. Go to edit the phone number of contact information in the Setting page
   - In the verification code inputs, check if Arrow-Left, Arrow-Right and Backspace keyboards can move the input cursor as expected
5. Go to the Product Feed page
   - Hover to the edit icon on the Product Feed table
   - Check if the tooltip is shown and centered above the icon
   ![image](https://github.com/user-attachments/assets/1d49045a-f3aa-4c5b-89c9-224346ee467e)
6. View the e2e test result on GitHub Actions: https://github.com/woocommerce/google-listings-and-ads/actions/runs/10452211003/job/28940184227
7. Do some smoke testing

### Changelog entry

> Tweak - Make the Tooltip use the new placement prop when WordPress >= 6.4.
> Tweak - Replace deprecated event.keyCode with event.code for the verification code inputs in the contact information setting.
> Dev - Migrate jest tests to use Node.js 20.
> Dev - Avoid accidentally using the event object to reset the asset group values in the CampaignAssetsForm component.
> Dev - Rewrite the replacer of JSON.stringify in getReportKey to ensure it returns the same key regardless of the query keys' order.
